### PR TITLE
1/n Simplify spawn plugins: Simplify handling of multiprocessing queue

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #10059, implements the first 3 steps which are highly coupled. Please read steps 1-3 on the issue to better follow the thought process behind this PR. Steps 1-3 are combined in this PR because they are tightly entangled. Steps 4 and 5 will be independent follow-ups.

Notes to reviewer:
- TPUSpawnPlugin subclasses from DDPSpawnPlugin and it is therefore not very feasible to split this PR into two pieces. Due to the inheritance, it makes the most sense to update both plugins in the same PR.
- Please have a look at #10865 if you find the time.

### Does your PR introduce any breaking changes? If yes, please list them.

A very tiny edge case breaking change in two hooks, `LightningModule.add_to_queue` and `LightningModule.get_from_queue` which now accept a `_SimpleQueue` as input instead of a `torch.multiprocessing.SimpleQueue`. This is however not expected to make an impact directly, since only the type hint changes while the user interface remains exactly the same.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
   _No new tests are required. We have enough (quite expensive) ddp spawn tests that will pass and confirm that these refactors and redesigns retain the existing functionality and performance._
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @justusschock @awaelchli @akihironitta @tchaton @kaushikb11

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 1]
```diff
@@ -208,28 +199,11 @@ def new_process(self, trainer: "pl.Trainer", mp_queue: SimpleQueue) -> None:
...
 
         # ensure that spawned processes go through teardown before joining
         trainer._call_teardown_hook()
-
-    def post_dispatch(self, trainer: "pl.Trainer"):
-        # restore main state with best weights
-        best_path = self.mp_queue.get()
-        last_path = self.mp_queue.get()
-        self._results = self.mp_queue.get()
-        # get the `callback_metrics` and set it to the trainer
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:223]: Reviewer: these get calls (and corresponding put calls) are no longer needed. The core of this PR is to move the whole tuple of results all at once. This now happens centrally inside the `spawn()` method, aligning both Lite and Lightning.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 2]
```diff
@@ -208,28 +199,11 @@ def new_process(self, trainer: "pl.Trainer", mp_queue: SimpleQueue) -> None:
...
-        # only in case the user does not override it.
-        # TODO: Remove the if in v1.7
-        if is_overridden("get_from_queue", self.lightning_module):
-            self.lightning_module.get_from_queue(self.mp_queue)
-        else:
-            self.get_from_queue(trainer, self.mp_queue)
-
-        # recover the weights of the processes trained in the children
-        self.__recover_child_process_weights(best_path, last_path)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:232]: Reviewer: this has moved to `__recover_results_in_main_process`

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 3]
```diff
@@ -413,3 +400,16 @@ def teardown(self) -> None:
...
+    """Simulates a :class:`torch.multiprocessing.queue.SimpleQueue` interface using the Python list."""
+
+    def get(self) -> Any:
+        return self.pop(0)
+
+    def put(self, item: Any) -> None:
+        self.append(item)
+
+    def empty(self) -> bool:
+        return len(self) == 0
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:415]: Reviewer: This is needed to preserve backward compatibility with everyone overriding the `get_form_queue` and `add_to_queue` hooks expecting a queue object. See #10865 for how I propose to go forward.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 4]
```diff
@@ -290,18 +289,6 @@ def lightning_module(self) -> Optional["pl.LightningModule"]:
         """Returns the pure LightningModule without potential wrappers."""
         return unwrap_lightning_module(self._model) if self._model is not None else None
 
-    @property
-    def results(self) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
-        """Enables plugin-agnostic access to the result returned by the training/evaluation/prediction run.
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:295]: Reviewer: I can introduce a deprecation if you request it.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 5]
```diff
@@ -285,18 +261,19 @@ def __transfer_distrib_spawn_state_on_fit_end(self, trainer: "pl.Trainer", resul
...
+        extra = _ExtraQueue()
         if is_overridden("add_to_queue", self.lightning_module):
-            self.lightning_module.add_to_queue(self.mp_queue)
+            # TODO: Remove the if in v1.7
+            self.lightning_module.add_to_queue(extra)
         else:
-            self.add_to_queue(trainer, self.mp_queue)
+            self.add_to_queue(trainer, extra)
+
+        return best_model_path, last_path, results, extra
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:272]: If this is not going to be removed soon in a follow-up, I'd suggest a NamedTuple here, it'll aid with the typing too

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 6]
```diff
@@ -413,3 +400,16 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+
+class _ExtraQueue(list):
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:405]:
```suggestion
class _MockQueue(list):
```

or

```suggestion
class _FakeQueue(list):
```

**Review by carmocca — APPROVED** [review]
Nice!

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1917,7 +1917,7 @@ def model_size(self) -> float:
             )
         return get_model_size_mb(self)
 
-    def add_to_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
+    def add_to_queue(self, queue: pl.plugins.training_type.ddp_spawn._ExtraQueue) -> None:
```
Comment (four4fish) [inline pytorch_lightning/core/lightning.py:1920]: Curious why do we have this in lightningModule? Can we call directly from DDPSpawnPlugin?

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:272]: Agreed, NamedTuple/TypedDict/Dataclass will make this more explicit

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (four4fish) [reply pytorch_lightning/core/lightning.py:1920]: nvm!

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 8]
```diff
@@ -145,16 +132,22 @@ def set_world_ranks(self, process_idx: int = 0) -> None:
     def get_mp_spawn_kwargs(self, trainer: Optional["pl.Trainer"] = None) -> Dict[str, Any]:
```
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:132]: Do we still need this function?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 9]
```diff
@@ -169,9 +164,7 @@ def init_dist_connection(self, global_rank: int, world_size: int) -> None:
     def set_world_ranks(self, process_idx: int = 0) -> None:
         pass
 
-    def new_process(self, trainer: "pl.Trainer", mp_queue: SimpleQueue) -> None:
-        self.mp_queue = mp_queue
-
+    def new_process(self, trainer: "pl.Trainer") -> Optional[Tuple[Optional[str], Optional[str], Any, "_ExtraQueue"]]:
```
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:167]: In ddp (non-spawning plugins), we only need pass trainer to strategy for setup() which calls init_optimizer. After we clean up TrainerOptimizersMixin, most plugin doesn't need trainer anymore. 
Do you think it's possible to spawn individual functions instead of entire trainer for spawning in the furture?

**Comment by four4fish** [comment]
This is super cool!! love it!

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:272]: Ah true, I was thinking about this too. 
Any of the three options containers would be ok I guess, as long as it's pickleable. 
What's the preference?

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:405]: I agree with _FakeQueue

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:132]: The original motivation for it to be a standalone method is that users/authors of plugins can just override only this part if they need to. If we remove it, they would instead override the spawn() method directly, or perhaps we could expose the `start_method` as an attribute to the plugin. Let's discuss it in an issue.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:272]: +1 namedTuple

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:167]: In step 4 and 5 of  #10059 we will automatically get rid of some of these trainer references yes. There are some occurrences in the plugins where we need to check things like `trainer.stage == FITTING` but these trainer instances can also be accessed through model.trainer if people think this is better.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 10]
```diff
@@ -101,15 +100,6 @@ def num_nodes(self, num_nodes: int) -> None:
     def local_rank(self) -> int:
         return self._local_rank
 
-    def __getstate__(self):
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:104]: Neat !

**Review by tchaton — APPROVED** [review]
Looks great! Awesome work.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 11]
```diff
@@ -413,3 +398,23 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+
+class _FakeQueue(list):
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:403]: afaik subclassing mutable builtins (especially dict and list) can sometimes lead to unexpected results. Maybe use [`UserList`](https://docs.python.org/3/library/collections.html#collections.UserList) as base class instead?

**Review by justusschock — APPROVED** [review]
LGTM. Just one minor comment :)

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py]: going with namedTuple for now

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:133]: #10895

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 12]
```diff
@@ -161,9 +156,7 @@ def init_dist_connection(self, global_rank: int, world_size: int) -> None:
     def set_world_ranks(self, process_idx: int = 0) -> None:
         pass
 
-    def new_process(self, trainer: "pl.Trainer", mp_queue: SimpleQueue) -> None:
-        self.mp_queue = mp_queue
-
+    def new_process(self, trainer: "pl.Trainer") -> Optional[Tuple[Optional[str], Optional[str], Any, "_FakeQueue"]]:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:159]: Isn't this missing the SpawnOutput return?

File: pytorch_lightning/plugins/training_type/sharded_spawn.py [hunk 13]
```diff
@@ -115,12 +114,12 @@ def pre_backward(self, closure_loss: torch.Tensor) -> None:
     def post_training_step(self):
         pass
 
-    def new_process(self, trainer: "pl.Trainer", mp_queue: SimpleQueue) -> None:
+    def new_process(self, trainer: "pl.Trainer") -> Optional[Tuple[Optional[str], Optional[str], Any, _FakeQueue]]:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/sharded_spawn.py:117]: Same comment

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 11 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py]: @justusschock I don't think that's true nowadays: https://stackoverflow.com/questions/25464647/list-vs-userlist-and-dict-vs-userdict

`list` is the recommended approach for simple extension
and `collections.*` for complex extension

File: pytorch_lightning/plugins/training_type/sharded_spawn.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/sharded_spawn.py:117]: true, I forgot to update the subclasses. Fortunately new_process will become obsolete in 2/n if everything goes according to plan.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 11 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py]: Quoting Raymond Hettinger (Python core dev):

"FWIW, the UserDict class in Py3.x was relocated to the collections
module.  It isn't gone.  We made an effort to kill it but found that
there were compelling use cases that were not a cleanly solved by any
other approach."

https://bugs.python.org/issue5402

File: pytorch_lightning/plugins/training_type/sharded_spawn.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/sharded_spawn.py:117]: resolved in #10934
